x-pack/plugins/apm/server/utils/source_map_rt.ts

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { either } from 'fp-ts/lib/Either';
+import * as t from 'io-ts';
+
+const sourceMapRt = t.intersection([
+  t.type({
+    version: t.number,
+    sources: t.array(t.string),
+    mappings: t.string,
+  }),
+  t.partial({
+    names: t.array(t.string),
+    file: t.string,
+    sourceRoot: t.string,
+    sourcesContent: t.array(t.string),
+  }),
+]);
+
+export type SourceMap = t.TypeOf<typeof sourceMapRt>;
+
+export const toSourceMapRt = new t.Type<SourceMap, Buffer | string, unknown>(
+  'toSourceMapRt',
+  (u): u is SourceMap => u !== undefined,
+  (input, context) => {
+    return either.chain(
+      t.unknown.validate(input, context),
```

Comment:
@dgieselaar I'm not sure if this would be the correct implementation, would you know? I need the input to be either a Buffer or a string

Reply:
It's a buffer because I'm uploading a file.

```
curl -H "Content-Type: multipart/form-data" -H "kbn-xsrf: true" http://localhost:5601/foo/api/apm/sourcemaps -F sourcemap=@bundle.js.map -F service_name=bar -F service_version=5.0.0 -F bundle_filepath=/a/b/c
```

Reply:
how do we deal with encoding? do we expect it to always be utf-8?

Reply:
I think so, maybe @axw can chime in.

Reply:
According to https://sourcemaps.info/spec.html:

> For simplicity, the character set encoding is always UTF-8.

In theory a user might have some other _transfer_ encoding (Content-Transfer-Encoding), but we don't support this in APM Server at the moment so I think it's fine to assume it's transferred as-is.

Reply:
sgtm, @cauemarcondes do you know if it's possible to throw an error if the encoding is _not_ utf-8?

Reply:
I didn't find a way to thrown an error if not utf-8 @dgieselaar